File: src/main/java/com/ragin/bdd/cucumber/matcher/ValidDateMatcher.java
package com.ragin.bdd.cucumber.matcher;

import com.ragin.bdd.cucumber.datetimeformat.BddCucumberDateTimeFormat;
import com.ragin.bdd.cucumber.utils.DateUtils;
import java.util.Collection;
import lombok.RequiredArgsConstructor;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.springframework.stereotype.Component;

/**
 * Matcher for valid date
 */
@Component
@RequiredArgsConstructor
public class ValidDateMatcher extends BaseMatcher<Object> {
    private final Collection<BddCucumberDateTimeFormat> dateTimeFormatCollection;

    public boolean matches(Object item) {
        return DateUtils.isValidMandatoryDate(item, dateTimeFormatCollection);
    }

    @Override
    public void describeMismatch(Object item, Description description) {
        description.appendText("Is not a valid date: ").appendValue(item);
    }

    @Override
    public void describeTo(Description description) {
        // nothing to describe here
    }
}

File: src/main/kotlin/com/ragin/bdd/cucumber/matcher/ValidDateMatcher.kt
package com.ragin.bdd.cucumber.matcher

import com.ragin.bdd.cucumber.datetimeformat.BddCucumberDateTimeFormat
import com.ragin.bdd.cucumber.utils.DateUtils
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.springframework.stereotype.Component

/**
 * Matcher for valid date.
 *
 * ${json-unit.matches:isValidDate}
 */
@Component
class ValidDateMatcher(
    private val dateTimeFormatCollection: Collection<BddCucumberDateTimeFormat>
) : BaseMatcher<Any>(), BddCucumberJsonMatcher {
    override fun matches(item: Any): Boolean {
        return DateUtils.isValidMandatoryDate(
            dateObject = item,
            bddDateTimeFormats = dateTimeFormatCollection
        )
    }

    override fun describeMismatch(item: Any, description: Description) {
        description.appendText("Is not a valid date: ").appendValue(item)
    }

    override fun describeTo(description: Description) {
        // nothing to describe here
    }

    override fun matcherName(): String {
        return "isValidDate"
    }

    override fun matcherClass(): Class<out BaseMatcher<*>> {
        return this::class.java
    }
}
